File: App/frontend/src/store/searchMemorySettings.ts
export type EmbeddingProviderType = string;
export type SearchMemoryTarget = 'search' | 'memory';

export interface EmbeddingConfig {
  provider: EmbeddingProviderType;
  model: string;
  dimensions: number | null;
}

export interface RetrievalConfig {
  topKPerQuery: number;
  neighborWindow: number;
  maxPrimaryItems: number;
  maxTotalItems: number;
}

export interface SearchGeneralConfig {
  embedding: EmbeddingConfig;
  retrieval: RetrievalConfig;
  keywordPageSize: number;
}

export interface MemoryConfig {
  embedding: EmbeddingConfig;
  retrieval: RetrievalConfig;
}

export interface SearchMemorySettings {
  enabled: boolean;
  general: SearchGeneralConfig;
  overrides: {
    search?: SearchGeneralConfig;
    memory?: MemoryConfig;
  };
}

const DEFAULT_EMBEDDING: EmbeddingConfig = {
  provider: 'openai',
  model: '',
  dimensions: null,
};

const DEFAULT_RETRIEVAL: RetrievalConfig = {
  topKPerQuery: 20,
  neighborWindow: 0,
  maxPrimaryItems: 20,
  maxTotalItems: 60,
};

export function makeInitialSearchMemorySettings(): SearchMemorySettings {
  return {
    enabled: false,
    general: {
      embedding: { ...DEFAULT_EMBEDDING },
      retrieval: { ...DEFAULT_RETRIEVAL },
      keywordPageSize: 20,
    },
    overrides: {},
  };
}

function cloneValue<T>(value: T): T {
  if (value === null || value === undefined) return value;
  if (Array.isArray(value)) return value.map((item) => cloneValue(item)) as T;
  if (typeof value === 'object') {
    return Object.fromEntries(
      Object.entries(value as Record<string, unknown>).map(([key, child]) => [key, cloneValue(child)])
    ) as T;
  }
  return value;
}

function normalizeEmbeddingConfig(config?: Partial<EmbeddingConfig> | null): EmbeddingConfig {
  return {
    provider: (config?.provider ?? DEFAULT_EMBEDDING.provider) as EmbeddingProviderType,
    model: typeof config?.model === 'string' ? config.model : DEFAULT_EMBEDDING.model,
    dimensions: typeof config?.dimensions === 'number' ? config.dimensions : null,
  };
}

function normalizeRetrievalConfig(config?: Partial<RetrievalConfig> | null): RetrievalConfig {
  const normalized: RetrievalConfig = {
    topKPerQuery: typeof config?.topKPerQuery === 'number' ? config.topKPerQuery : DEFAULT_RETRIEVAL.topKPerQuery,
    neighborWindow: typeof config?.neighborWindow === 'number' ? config.neighborWindow : DEFAULT_RETRIEVAL.neighborWindow,
    maxPrimaryItems: typeof config?.maxPrimaryItems === 'number' ? config.maxPrimaryItems : DEFAULT_RETRIEVAL.maxPrimaryItems,
    maxTotalItems: typeof config?.maxTotalItems === 'number' ? config.maxTotalItems : DEFAULT_RETRIEVAL.maxTotalItems,
  };
  if (normalized.maxTotalItems < normalized.maxPrimaryItems) {
    normalized.maxTotalItems = normalized.maxPrimaryItems;
  }
  return normalized;
}

function normalizeSearchGeneralConfig(config?: Partial<SearchGeneralConfig> | null): SearchGeneralConfig {
  return {
    embedding: normalizeEmbeddingConfig(config?.embedding),
    retrieval: normalizeRetrievalConfig(config?.retrieval),
    keywordPageSize: typeof config?.keywordPageSize === 'number' ? config.keywordPageSize : 20,
  };
}

function normalizeMemoryConfig(config?: Partial<MemoryConfig> | null): MemoryConfig {
  return {
    embedding: normalizeEmbeddingConfig(config?.embedding),
    retrieval: normalizeRetrievalConfig(config?.retrieval),
  };
}

function normalizeSearchMemorySettings(settings: SearchMemorySettings): SearchMemorySettings {
  const normalized: SearchMemorySettings = {
    enabled: Boolean(settings?.enabled),
    general: normalizeSearchGeneralConfig(settings?.general),
    overrides: {},
  };

  if (settings?.overrides?.search) {
    normalized.overrides.search = normalizeSearchGeneralConfig(settings.overrides.search);
  }
  if (settings?.overrides?.memory) {
    normalized.overrides.memory = normalizeMemoryConfig(settings.overrides.memory);
  }

  return normalized;
}

function resolveSearchSettingsFromNormalized(settings: SearchMemorySettings): SearchGeneralConfig {
  return cloneValue(settings.overrides.search ?? settings.general);
}

function resolveMemorySettingsFromNormalized(settings: SearchMemorySettings): MemoryConfig {
  return cloneValue(
    settings.overrides.memory ?? {
      embedding: settings.general.embedding,
      retrieval: settings.general.retrieval,
    }
  );
}

export function validateSearchMemorySettings(settings: SearchMemorySettings): SearchMemorySettings {
  const normalized = normalizeSearchMemorySettings(settings);

  if (normalized.enabled) {
    const checks = [
      ['search', resolveSearchSettingsFromNormalized(normalized).embedding],
      ['memory', resolveMemorySettingsFromNormalized(normalized).embedding],
    ] as const;
    for (const [, embedding] of checks) {
      if (!embedding.provider.trim() || !embedding.model.trim()) {
        throw new Error('Search & Memory embedding provider/model is required when enabled');
      }
    }
  }

  return normalized;
}

export function hasSearchMemoryOverride(settings: SearchMemorySettings, target: SearchMemoryTarget): boolean {
  const normalized = normalizeSearchMemorySettings(settings);
  return Object.prototype.hasOwnProperty.call(normalized.overrides, target);
}

export function makeSearchOverrideSeed(settings: SearchMemorySettings): SearchGeneralConfig {
  return cloneValue(normalizeSearchMemorySettings(settings).general);
}

export function makeMemoryOverrideSeed(settings: SearchMemorySettings): MemoryConfig {
  const normalized = normalizeSearchMemorySettings(settings);
  return {
    embedding: cloneValue(normalized.general.embedding),
    retrieval: cloneValue(normalized.general.retrieval),
  };
}

export function resolveSearchSettings(settings: SearchMemorySettings): SearchGeneralConfig {
  const normalized = normalizeSearchMemorySettings(settings);
  return resolveSearchSettingsFromNormalized(normalized);
}

export function resolveMemorySettings(settings: SearchMemorySettings): MemoryConfig {
  const normalized = normalizeSearchMemorySettings(settings);
  return resolveMemorySettingsFromNormalized(normalized);
}

File: App/backend/services/search_memory_settings.py
from __future__ import annotations

from copy import deepcopy
from typing import Any, Literal

from pydantic import ValidationError

from ..provider_engine.registry import require_provider
from ..schemas.settings import (
    MemoryConfig,
    SearchGeneralConfig,
    SearchMemorySettings,
)


SearchMemoryTarget = Literal["search", "memory"]


INITIAL_SEARCH_MEMORY_SETTINGS: dict[str, Any] = {
    "enabled": False,
    "general": {
        "embedding": {
            "provider": "openai",
            "model": "",
            "dimensions": None,
        },
        "retrieval": {
            "topKPerQuery": 20,
            "neighborWindow": 0,
            "maxPrimaryItems": 20,
            "maxTotalItems": 60,
        },
        "keywordPageSize": 20,
    },
    "overrides": {},
}


def make_initial_search_memory_settings() -> dict[str, Any]:
    return deepcopy(INITIAL_SEARCH_MEMORY_SETTINGS)


def _normalize_search_memory_settings(search_memory_settings: dict[str, Any]) -> dict[str, Any]:
    if not isinstance(search_memory_settings, dict):
        raise ValueError("searchMemorySettings must be an object")

    try:
        validated = SearchMemorySettings.model_validate(search_memory_settings)
    except ValidationError as exc:
        raise ValueError(str(exc)) from exc

    normalized = validated.model_dump(mode="json")
    if not isinstance(normalized.get("overrides"), dict):
        normalized["overrides"] = {}
    return normalized


def _normalize_general_config(config: dict[str, Any]) -> dict[str, Any]:
    validated = SearchGeneralConfig.model_validate(config)
    return validated.model_dump(mode="json")


def _normalize_memory_config(config: dict[str, Any]) -> dict[str, Any]:
    validated = MemoryConfig.model_validate(config)
    return validated.model_dump(mode="json")


def make_search_override_seed(settings: dict[str, Any]) -> dict[str, Any]:
    normalized = validate_search_memory_settings(settings)
    return deepcopy(normalized["general"])


def make_memory_override_seed(settings: dict[str, Any]) -> dict[str, Any]:
    normalized = validate_search_memory_settings(settings)
    general = normalized["general"]
    return {
        "embedding": deepcopy(general["embedding"]),
        "retrieval": deepcopy(general["retrieval"]),
    }


def has_search_memory_override(settings: dict[str, Any], target: SearchMemoryTarget) -> bool:
    normalized = validate_search_memory_settings(settings)
    overrides = normalized.get("overrides", {})
    return isinstance(overrides, dict) and isinstance(overrides.get(target), dict)


def _resolve_search_settings_from_validated(settings: dict[str, Any]) -> dict[str, Any]:
    override = settings["overrides"].get("search")
    if isinstance(override, dict):
        return deepcopy(_normalize_general_config(override))
    return deepcopy(settings["general"])


def resolve_search_settings(settings: dict[str, Any]) -> dict[str, Any]:
    normalized = validate_search_memory_settings(settings)
    return _resolve_search_settings_from_validated(normalized)


def _resolve_memory_settings_from_validated(settings: dict[str, Any]) -> dict[str, Any]:
    override = settings["overrides"].get("memory")
    if isinstance(override, dict):
        return deepcopy(_normalize_memory_config(override))
    general = settings["general"]
    return {
        "embedding": deepcopy(general["embedding"]),
        "retrieval": deepcopy(general["retrieval"]),
    }


def resolve_memory_settings(settings: dict[str, Any]) -> dict[str, Any]:
    normalized = validate_search_memory_settings(settings)
    return _resolve_memory_settings_from_validated(normalized)


def validate_search_memory_settings(search_memory_settings: dict[str, Any]) -> dict[str, Any]:
    normalized = _normalize_search_memory_settings(search_memory_settings)

    if normalized.get("enabled"):
        search = _resolve_search_settings_from_validated(normalized)
        memory = _resolve_memory_settings_from_validated(normalized)
        checks = (
            ("Search", search["embedding"]),
            ("Memory", memory["embedding"]),
        )
        for label, embedding in checks:
            provider = str(embedding.get("provider") or "").strip()
            model = str(embedding.get("model") or "").strip()
            provider_spec = require_provider(provider)
            if provider_spec.embedding is None:
                raise ValueError(f"{label} embedding provider '{provider}' is not supported")
            if not provider or not model:
                raise ValueError(f"{label} embedding provider/model is required when Search & Memory is enabled")

    return normalized


def clear_dimensions_for_target(settings: dict[str, Any], target: SearchMemoryTarget) -> dict[str, Any]:
    normalized = validate_search_memory_settings(settings)
    overrides = normalized.setdefault("overrides", {})
    if target == "search" and isinstance(overrides.get("search"), dict):
        overrides["search"]["embedding"]["dimensions"] = None
        return normalized
    if target == "memory" and isinstance(overrides.get("memory"), dict):
        overrides["memory"]["embedding"]["dimensions"] = None
        return normalized
    normalized["general"]["embedding"]["dimensions"] = None
    return normalized


__all__ = [
    "SearchMemoryTarget",
    "clear_dimensions_for_target",
    "has_search_memory_override",
    "make_initial_search_memory_settings",
    "make_memory_override_seed",
    "make_search_override_seed",
    "resolve_memory_settings",
    "resolve_search_settings",
    "validate_search_memory_settings",
]
